Classification surrogates

gmancino opened the pull request:
# Add Classification Models for Surrogates

Adding new surrogates to allow for classification of output values (e.g. 'unacceptable', 'acceptable', or 'ideal') to use for modeling unknown constraints. Concretely, if $g_{\theta}:\mathbb{R}^d\to[0,1]^c$ represents a function governed by learnable parameters $\theta$ which outputs a probability vector over $c$ potential classes (i.e. for input $x\in\mathbb{R}^d$, $g_{\theta}(x)^\top\mathbf{1}=1$ where $\mathbf{1}$ is the vector of all 1's) and we have acceptability criteria for the corresponding classes given by $a\in\{0,1\}^c$, we can compute the expected acceptability as scalar output via $g_{\theta}(x)^\top a\in[0,1]$ which can be passed in as a constrained objective function.

### Classification Models

We add a new objective function in 'bofire/data_models/objectives/categorical.py' which can be passed to outputs of type `CategoricalOutput`. These are instantiated with a list of probability scales (i.e. the acceptability criteria vector) via the `desirability` argument and inherit the categories from the corresponding output. Using this new objective:

1. We implement an MLP ensemble method to start ('bofire/data_models/surrogates/mlp.py') which outputs a $c$-dimensional probability vector for each datapoint
2. The predicted value (stored in `{key}_{class}_prob` of the predictions) is the argmax along this probability vector, while the objective value (stored in `{key}_{class}_{des}` of the predictions) is the inner-product of the probability vector with the acceptability criteria vector
    - This value is also computed in the `constrained_objective2botorch` function, which currently undergoes the inverse of the sigmoid transformation to maintain the value in the probability space
3. We pass in the objective value to BoTorch as a constraint

gmancino commented:
@jduerholt, please let me know if the **style** of the updates here are appropriate. I will build more models if the initial idea makes sense.

jduerholt commented on bofire/strategies/doe/utils_categorical_discrete.py:
Why changes in this rather unrelated module?

gmancino replied:
Ruff was failing - something about using isinstance is better than comparing type

jduerholt commented on bofire/data_models/features/categorical.py:
```diff
@@ -360,37 +359,43 @@ class CategoricalOutput(Output):
-    objective: Annotated[
-        List[Annotated[float, Field(type=float, ge=0, le=1)]], Field(min_items=2)
-    ]
+    objective: CategoricalObjective = Field(
```
I do not understand this change? Why the need to model the desirability of the individual categories by a an own class? Why not keeping the list (ideally transforming it to a tuple?)

gmancino replied:
The challenge is that we need to pass the `CategoricalOutput` type to the `constrained_objective2botorch` to be evaluated by BoTorch as a constraint. To me, this seemed like a more reasonable option as it did not require refactoring the `constrained_objective2botorch` to also look for lists (tuples really). In the end, the objective should produce a scalar of some kind and having the objective just as a list does not allow for such functionality. Maybe I am missunderstanding the question...

jduerholt replied:
Ok, makes sense. Then let's do it as proposed by you by an own objective class which holds a tuple of desirability values which are between 0 and 1.

jduerholt commented on bofire/utils/torch_tools.py:
```diff
@@ -147,24 +148,50 @@ def constrained_objective2botorch(
+        return (
+            [
+                lambda Z: -1.0
+                * objective.w
```
Why `objective.w`?

gmancino replied:
This is an error. I was testing things locally; it will be changed in the next revision :)

jduerholt commented on bofire/data_models/features/categorical.py:
```diff
@@ -360,37 +359,43 @@ class CategoricalOutput(Output):
-                raise ValueError("Objective values has to be smaller equal than 1.")
-            if o < 0:
-                raise ValueError("Objective values has to be larger equal than zero")
+        for w in weights:
```
If we go for this structure, this validation should go to the inside of the objective class as validator there, or we just do it via the annotated stuff. But then also there.

jduerholt commented on bofire/data_models/features/categorical.py:
```diff
@@ -360,37 +359,43 @@ class CategoricalOutput(Output):
+            if w < 0:
+                raise ValueError("Objective weight has to be larger equal than zero")
+        # Save the categories to the objective if they do not exist
+        objective.categories = (
```
Hmm, why is this needed? This is not a good data structure from a software engineering point of view.

gmancino replied:
Is this referring to the reassignment of the `objective.categories`? I can change this so that `CategoricalObjective` takes in both objective and categories arguments. Then the validator just makes sure the categories are the same. Does that seem better?

Initially, I thought the repetition of categories was redundant, but maybe it makes more sense?

jduerholt replied:
I meant that the `CategoricalObjective` should not hold the categories, it should only hold the `desirabilities` or `weights` and the categories should just be attributes of the `CategoricalOutput.`

gmancino replied:
I see - I have updated accordingly

gmancino replied:
Just coming back to this; is there a reason the `CategoricalObjective` should not hold the categories? I think there is an inherit challenge with decoupling the `CategoricalOutput` and `CategoricalObjective`, but both functionalities serve different purposes in BoFire...

It might make sense to have the objective hold both the categories and the desireabilities, that way the CategoricalOutput can inherit those values and do any text processing/whatever else is done with Outputs and the Objective can handle the actual computations

jduerholt commented on bofire/surrogates/mlp_classifier.py:
```diff
@@ -0,0 +1,220 @@
+            X: A `batch_shape x n x d`-dim input tensor `X`.
+
+        Returns:
+            A `batch_shape x s x n x m x C`-dimensional output tensor where
```
and what is n and m?

gmancino replied:
Copy paste typo - I will fix these as it should just be `batch_size x n x C`

jduerholt commented on bofire/surrogates/mlp_classifier.py:
```diff
@@ -0,0 +1,220 @@
+            A `batch_shape x s x n x m x C`-dimensional output tensor where
+            `s` is the size of the ensemble and `C` is the number of classes.
+        """
+        return torch.stack([torch.softmax(mlp(X), dim=-1) for mlp in self.mlps], dim=-3)
```
Why do we softmax again the result of the log_softmax?

gmancino replied:
Error on my part - will change to `torch.stack([mlp(X).exp() for mlp in self.mlps], dim=-3)`

jduerholt reviewed (commented):
Hi Gabe, this looks really impressive. I just added some intitial questions as I am not yet fully through this beast :D

jduerholt commented on bofire/utils/torch_tools.py:
```diff
@@ -147,24 +148,50 @@ def constrained_objective2botorch(
-        Tuple[List[Callable[[Tensor], Tensor]], List[float]]: List of callables that can be used by botorch for setting up the constrained objective, and
-            list of the corresponding botorch eta values.
+        Tuple[List[Callable[[Tensor], Tensor]], List[float], int]: List of callables that can be used by botorch for setting up the constrained objective,
+            list of the corresponding botorch eta values, final index used by the method (to track for categorical variables)
```
For what is this index needed?

gmancino replied:
Since the MLP will output a vector of length `c`, we need to keep track of two items: 1. an index related to the current objective and 2. an index related to the current objective's output. For most other methods, if the output is 1 dimensional, then these two indexes are the same. For the any method that outputs more than one objective value, then eventually these two indexes will be misaligned. 

The output that eventually gets passed to `constrained_objective2botorch` is the output from model.posterior <- which is the result of the forward pass through all models. If one of these models outputs a vector of size > 1, then using solely the objective index results in a failure to capture the underlying model's output. Is this a problem with other methods too? Do I need to add a check for all models somewhere else to measure their output size?

jduerholt replied:
Ah, I get it. This makes sense. Currently, I do not see any problem there, but I will think about it ;)

gmancino commented:
@jduerholt thank you so much for your feedback! I think I have addressed all of your concerns, but please let me know if there are any additional errors. If it all looks good, I plan on implementing something like in https://github.com/pytorch/botorch/issues/640 for a categorical GP. This should cover our initial basis :)

jduerholt commented on bofire/utils/torch_tools.py:
```diff
@@ -147,24 +148,49 @@ def constrained_objective2botorch(
+                * (
+                    Z[..., idx : idx + len(objective.desirability)]
+                    * torch.tensor(objective.desirability).to(**tkwargs)
+                ).sum(-1)
```
But maybe, I overlooked a transformation somewhere.

gmancino commented on bofire/data_models/features/categorical.py:
```diff
@@ -401,9 +404,33 @@ def validate_experimental(self, values: pd.Series) -> pd.Series:
     def to_dict(self) -> Dict:
         """Returns the catergories and corresponding objective values as dictionary"""
-        return dict(zip(self.categories, self.objective))
+        return dict(zip(self.categories, self.objective.desirability))
```
This would mean we set up functions within the objective which take in the categories and do some processing. I think it's still unclear to me why the categories cannot be directly passed to the objective upon instantiation

gmancino commented on bofire/strategies/predictives/predictive.py:
```diff
@@ -102,16 +103,47 @@ def predict(self, experiments: pd.DataFrame) -> pd.DataFrame:
+                *[
+                    [f"{feat.key}_pred"]
+                    if not isinstance(feat, CategoricalOutput)
+                    else [f"{feat.key}_pred_{cat}" for cat in feat.categories]
```
This can be added. One concern is then the user cannot see the other probabilities. For example, it could be [0.51, 0.49, 0.0] as the predicted probabilities, but if they only see the highest value the information on the closeness of the first two classes is missed. I'm fine with either way, but just want to bring a scenario like this up

gmancino replied:
@jduerholt any comment on the above? Which functionality do you prefer?

gmancino commented:
Hi @jduerholt, I have made the corresponding changes to this PR which we have discussed. Some tests are currently failing because I changed the `TCategoryVals` to be of type `Tuple[str, ...]`, which would make the categories immutable. Hence, in each hard coded location of categories (which are currently lists) the tests fail. I know we discussed using Tuples instead of Lists for this so it may be worth changing by hand unless you have some other opinion? There may be additional technical comments on the changes made which need to be addressed first ;)

jduerholt commented on bofire/data_models/features/categorical.py:
```diff
@@ -14,14 +13,18 @@
     Attributes:
-        categories (List[str]): Names of the categories.
-        allowed (List[bool]): List of bools indicating if a category is allowed within the optimization.
+        categories (Tuple[str]): Names of the categories.
```
Let's do this in a later PR with changing the types here from list to tuple, as mentioned above. It could then even make sense to have it not as tuple but as set.

jduerholt commented on bofire/data_models/features/categorical.py:
```diff
@@ -363,39 +365,52 @@ class CategoricalOutput(Output):
-        List[Annotated[float, Field(type=float, ge=0, le=1)]], Field(min_items=2)
-    ]
+    objective: Optional[
+        Union[CategoricalObjective, ConstrainedCategoricalObjective]
```
Hmm, `CategoricalObjective` is an abstract Objective. We only allow here for the actual ones. You could introduce in `data_models/objectives/api.py` a type called `AnyCategoricalObjective`, which is for now just equal to ConstrainedCategoricalObjective.

jduerholt commented on bofire/data_models/features/categorical.py:
```diff
@@ -363,39 +365,52 @@ class CategoricalOutput(Output):
+        Returns:
+            Tuple[str]: Tuple of the categories
+        """
+        if objective.categories != tuple(values["categories"]):
```
We sould have a test for it within `tests/data_models/specs/features.py` using the `add_invalid` syntax.

jduerholt commented on bofire/data_models/features/categorical.py:
```diff
@@ -363,39 +365,52 @@ class CategoricalOutput(Output):
         return objective
 
+    @classmethod
+    def from_objective(
```
We need a test for this in `tests/bofire/data_models/features/categorical.py`

jduerholt commented on bofire/data_models/features/categorical.py:
```diff
@@ -404,9 +419,14 @@ def validate_experimental(self, values: pd.Series) -> pd.Series:
-        return values.map(self.to_dict()).astype(float)
+        if self.objective is None:
+            return pd.Series(
+                data=[np.nan for _ in range(len(values))],
```
we need a test for this method.

gmancino replied:
It will be in features/test_categorical.py

jduerholt commented on bofire/data_models/objectives/api.py:
```diff
@@ -25,6 +29,8 @@
 ]
 
 AnyConstraintObjective = Union[
+    CategoricalObjective,
```
CategoricalObjective is just the abstract class. It should not be listed here.

jduerholt commented on bofire/data_models/objectives/categorical.py:
```diff
@@ -0,0 +1,106 @@
+            raise ValueError("categories must be unique")
+        return tuple(categories)
+
+    def __call__(self, x: Union[pd.Series, np.ndarray]) -> Union[pd.Series, np.ndarray]:
```
Remove this method, it should stay as abstract objective.

jduerholt commented on bofire/data_models/objectives/categorical.py:
```diff
@@ -0,0 +1,106 @@
+            Tuple[bool]: Tuple of the desirability
+        """
+        if len(desirability) != len(values["categories"]):
+            raise ValueError(
```
We should add a test for this via the add_invalid syntax in `data_models/specs/objectives.py`

gmancino replied:
Done

jduerholt commented on bofire/data_models/objectives/categorical.py:
```diff
@@ -0,0 +1,106 @@
+            )
+        return tuple(desirability)
+
+    def to_dict(self) -> Dict:
```
We need tests for these methods.

gmancino replied:
They are in tests/bofire/data_models/domain/test_outputs.py. Let me know if they are insufficient or if you want them elsewhere (should be line 290)

jduerholt commented on bofire/data_models/surrogates/surrogate.py:
```diff
@@ -28,3 +29,17 @@ def validate_outputs(cls, v, values):
         if len(v) == 0:
             raise ValueError("At least one output feature has to be provided.")
         return v
+
```
We need the actual validator that uses the is_output_implemented method, have a look here: https://github.com/experimental-design/bofire/blob/73c1db7876d72e15694610e6b068126da96b98d5/bofire/data_models/strategies/strategy.py#L38 and then we need to test if for at least one surrogate using the `add_invalid` syntax in `specs/surrogates.py`

jduerholt commented on bofire/data_models/surrogates/fully_bayesian.py:
```diff
@@ -16,3 +17,16 @@ def validate_thinning(cls, value, values):
+        Returns:
+            bool: True if the output type is valid for the surrogate chosen, False otherwise
+        """
+        return True if isinstance(my_type, ContinuousOutput) else False
```
```suggestion
        return  isinstance(my_type, ContinuousOutput)
```

jduerholt commented on bofire/data_models/surrogates/linear.py:
```diff
@@ -14,3 +18,17 @@ class LinearSurrogate(TrainableBotorchSurrogate):
+        Returns:
+            bool: True if the output type is valid for the surrogate chosen, False otherwise
+        """
+        return True if isinstance(my_type, ContinuousOutput) else False
```
```suggestion
        return  isinstance(my_type, ContinuousOutput)
```

jduerholt commented on bofire/data_models/surrogates/mixed_single_task_gp.py:
```diff
@@ -29,3 +30,16 @@ def validate_categoricals(cls, v, values):
+        Returns:
+            bool: True if the output type is valid for the surrogate chosen, False otherwise
+        """
+        return True if isinstance(my_type, ContinuousOutput) else False
```
```suggestion
        return  isinstance(my_type, ContinuousOutput)
```

jduerholt commented on bofire/data_models/surrogates/mlp.py:
```diff
@@ -18,4 +19,17 @@ class MLPEnsemble(TrainableBotorchSurrogate):
+        Returns:
+            bool: True if the output type is valid for the surrogate chosen, False otherwise
+        """
+        return True if isinstance(my_type, ContinuousOutput) else False
```
```suggestion
        return  isinstance(my_type, ContinuousOutput)
```

jduerholt commented on bofire/data_models/surrogates/mlp_classifier.py:
```diff
@@ -0,0 +1,36 @@
+        Returns:
+            bool: True if the output type is valid for the surrogate chosen, False otherwise
+        """
+        return True if isinstance(my_type, CategoricalOutput) else False
```
```suggestion
        return  isinstance(my_type, CategoricalOutput)
```

jduerholt commented on bofire/data_models/surrogates/polynomial.py:
```diff
@@ -23,3 +24,16 @@ def from_power(power: int, inputs: Inputs, outputs: Outputs):
+        Returns:
+            bool: True if the output type is valid for the surrogate chosen, False otherwise
+        """
+        return True if isinstance(my_type, ContinuousOutput) else False
```
```suggestion
        return  isinstance(my_type, ContinuousOutput)
```

jduerholt commented on bofire/data_models/surrogates/random_forest.py:
```diff
@@ -30,5 +30,16 @@ class RandomForestSurrogate(TrainableBotorchSurrogate):
+        Returns:
+            bool: True if the output type is valid for the surrogate chosen, False otherwise
+        """
+        return True if isinstance(my_type, ContinuousOutput) else False
```
```suggestion
        return  isinstance(my_type, ContinuousOutput)
```

jduerholt commented on bofire/data_models/surrogates/single_task_gp.py:
```diff
@@ -108,3 +112,16 @@ class SingleTaskGPSurrogate(TrainableBotorchSurrogate):
+        Returns:
+            bool: True if the output type is valid for the surrogate chosen, False otherwise
+        """
+        return True if isinstance(my_type, ContinuousOutput) else False
```
```suggestion
        return  isinstance(my_type, ContinuousOutput)
```

jduerholt commented on bofire/data_models/surrogates/tanimoto_gp.py:
```diff
@@ -26,3 +27,16 @@ class TanimotoGPSurrogate(TrainableBotorchSurrogate):
+        Returns:
+            bool: True if the output type is valid for the surrogate chosen, False otherwise
+        """
+        return True if isinstance(my_type, ContinuousOutput) else False
```
```suggestion
        return  isinstance(my_type, ContinuousOutput)
```

jduerholt commented on bofire/data_models/surrogates/xgb.py:
```diff
@@ -75,3 +77,16 @@ def validate_input_preprocessing_specs(cls, v, values):
+        Returns:
+            bool: True if the output type is valid for the surrogate chosen, False otherwise
+        """
+        return True if isinstance(my_type, ContinuousOutput) else False
```
```suggestion
        return  isinstance(my_type, ContinuousOutput)
```

jduerholt reviewed (changes requested):
Hi Gabe,

I looked over the first part of the PR and let inline comments. I will look over the rest during the Christmas days. 

Main comments are due to missing tests.

Thanks!

Best,

Johannes

gmancino commented on bofire/surrogates/surrogate.py:
```diff
@@ -44,27 +45,80 @@ def predict(self, X: pd.DataFrame) -> pd.DataFrame:
+        pred_cols = []
+        sd_cols = []
+        for featkey in self.outputs.get_keys():
+            if hasattr(self.outputs.get_by_key(featkey), "categories"):
```
@jduerholt, there is a type error failing here. How do you recommend checking if the output type is categorical? Line 71 below also fails...

jduerholt replied:
`for feat in self.outputs.get(CategoricalOutput)`: Doing it like this, you only get categorical output features ;) The same filtereing also applies to `get_keys`. Have a look at the method in `domain.features.py`, they are super helpful and I use them really a lot, so you should familiarize with them.

jduerholt replied:
If you only need the keys use `get_keys(CategoricalOutput)`, it could be that you have still a type error, then add a `type: ignore`

gmancino commented on bofire/surrogates/mlp.py:
```diff
@@ -109,40 +125,50 @@ def forward(self, X: Tensor):
     lr: float = 1e-4,
     shuffle: bool = True,
     weight_decay: float = 0.0,
+    loss_function: Union[
```
@jduerholt there is a type error here. I tried removing the type from the loss function and also trying nn.module. I could try nn.modules.loss unless you have any other suggestions?

jduerholt replied:
Puh, no idea, then just set a `type: ignore` behind it ...

jduerholt commented:
Regarding you failing test, something went wrong in one of your merges against main, in main the line with the correct terms looks as following: https://github.com/experimental-design/bofire/blob/bf2a11999a7ba779a758c70ee17a555213676d6b/tests/bofire/strategies/doe/test_utils.py#L77

In your branch it looks likes this: https://github.com/experimental-design/bofire/blob/20ec136ffdf8bf662c9f332aac2f25bdfbf85116/tests/bofire/strategies/doe/test_utils.py#L77. In main we updated it at some point to the version looking like this `"x ** 2"` which is the new formulaic format, whereas you have in your branch still `x**2` which is the old format. 

I assume some problems with a merge.

jduerholt commented on tests/bofire/strategies/doe/test_utils.py:
```diff
@@ -74,7 +74,7 @@ def test_get_formula_from_string():
 
     # linear and quadratic
-    terms = ["1", "x0", "x1", "x2", "x0 ** 2", "x1 ** 2", "x2 ** 2"]
+    terms = ["1", "x0", "x1", "x2", "x0**2", "x1**2", "x2**2"]
```
One can see it also here that your branch differes from main in `doe/test_utils.py`

gmancino commented:
@jduerholt This is ready for another round of reviews :) Hopefully the last one!

gmancino commented on bofire/data_models/domain/features.py:
```diff
@@ -591,8 +596,8 @@ def get_by_objective(
                     filter_by_attribute(
-                        self.get(ContinuousOutput).features,
-                        lambda of: of.objective,  # type: ignore
+                        self.get([ContinuousOutput, CategoricalOutput]).features,
```
A test should be implemented in the appropriate file now.

gmancino commented on bofire/data_models/domain/features.py:
```diff
@@ -643,6 +652,14 @@ def __call__(
+                pd.Series(data=feat(experiments.filter(regex=f"{feat.key}_pred_")), name=f"{feat.key}_pred")  # type: ignore
+                if predictions
+                else experiments[feat.key]
+                for feat in self.features
```
Functionality should be updated to allow for `None` and solve the issues above.

gmancino commented on bofire/data_models/domain/features.py:
```diff
@@ -690,24 +707,24 @@ def validate_experiments(self, experiments: pd.DataFrame) -> pd.DataFrame:
                 [
-                    [f"{key}_pred", f"{key}_sd", f"{key}_des"]
-                    for key in self.get_keys_by_objective(Objective)
+                    [f"{obj.key}_pred", f"{obj.key}_sd", f"{obj.key}_des"]
```
Done

gmancino reviewed (commented):
I should have addressed all previous concerns. The tutorials also show the functionality of the new feature. We can add extra tests, but I added the major ones for now I believe.

jduerholt commented on bofire/surrogates/mlp.py:
```diff
@@ -109,40 +125,48 @@ def forward(self, X: Tensor):
     lr: float = 1e-4,
     shuffle: bool = True,
     weight_decay: float = 0.0,
+    loss_function=nn.L1Loss,  # type: ignore
```
Should we somehow list and check which we support currently?

gmancino replied:
How do you recommend?

gmancino replied:
Something like Union[torch.nn.modules.loss.L1Loss, ...]?

jduerholt replied:
Let's forget it. It is fine.

jduerholt commented on bofire/surrogates/mlp.py:
```diff
@@ -109,40 +125,48 @@ def forward(self, X: Tensor):
+            ):
                 targets = targets.reshape((targets.shape[0], 1))
+            else:
+                pass
```
An error message? Or we can just remove the else if we do nothing.

gmancino replied:
I will remove

jduerholt commented on bofire/data_models/surrogates/mlp.py:
```diff
@@ -18,4 +23,44 @@ class MLPEnsemble(TrainableBotorchSurrogate):
+
+class ClassificationMLPEnsemble(MLPEnsemble):
+    type: Literal["ClassificationMLPEnsemble"] = "ClassificationMLPEnsemble"
+    final_activation: Literal["softmax"] = "softmax"
```
`output_sclaer: Literal[ScalerEnum.Indetity] = ScalerEnum.Indetity`

I do not know if this works, but we have to forbid that a user assigns an output scaler, if it not works via the Literal, we have to add a field validator for this.

jduerholt commented on bofire/surrogates/surrogate.py:
```diff
@@ -44,27 +45,75 @@ def predict(self, X: pd.DataFrame) -> pd.DataFrame:
-            f"{key}_{t}" for key in self.outputs.get_keys() for t in ["pred", "sd"]
-        ]
+        expected_cols = []
+        check_columns = []
```
We are composing this kind of lists add a lof of different places, a few examples are:
- `data_models/domain/features.py`
- `data_models/strategies/predictives/predictive.py`

Can we create a utility method to create the correct col names etc. and move it into utils and just call it from different places. This will also making testing much easier.

jduerholt replied:
Let us put it into a new file under utils, as you want it ...

jduerholt commented on bofire/surrogates/trainable.py:
```diff
@@ -181,6 +182,22 @@ def cross_validate(
             y_train_pred = self.predict(X_train)  # type: ignore
+
+            # Convert to categorical if applicable
+            if isinstance(self.outputs.get_by_key(key).objective, ConstrainedCategoricalObjective):  # type: ignore
```
Hmm, why only of `ConstrainedCategoricalObjective`? This should apply always for every `CategoricalOutput` feature, or?

jduerholt replied:
And in general, I am not 100% sure, what should be achieved here.

gmancino replied:
CategoricalOutputs do not contain the functions to do the mapping (as per our previous discussions) from strings to categorical encodings. We need scalars for the loss function, not categories, for which Y would be an array of strings due to BoFire requiring that categorical columns be validated

jduerholt commented on bofire/utils/torch_tools.py:
```diff
@@ -188,24 +188,57 @@ def constrained_objective2botorch(
+                    )
+                )
+            ],
+            [objective.eta],
```
I would opt for removing the `objective.eta` and make sure that the eta which is given here is always 1 for identity, or?

gmancino replied:
I can do this, but will it break functionality elsewhere? Is eta always required for type Objective?

jduerholt replied:
Should not be a problem.

jduerholt commented on tests/bofire/utils/test_torch_tools.py:
```diff
@@ -739,7 +743,7 @@ def test_get_initial_conditions_generator(sequential: bool):
     ],
 )
 def test_constrained_objective2botorch(objective):
```
We need here also the tests for the new objective, or?

gmancino replied:
These are added :)

gmancino commented on bofire/data_models/domain/features.py:
```diff
@@ -690,24 +707,24 @@ def validate_experiments(self, experiments: pd.DataFrame) -> pd.DataFrame:
                     for key in self.get_keys_by_objective(
-                        excludes=Objective,
-                        includes=None,  # type: ignore
+                        excludes=Objective, includes=None  # type: ignore
```
Does it make sense for `CategoricalObjective` to be None type? What would you prefer?

gmancino commented on bofire/data_models/domain/features.py:
```diff
@@ -672,23 +689,18 @@ def validate_experiments(self, experiments: pd.DataFrame) -> pd.DataFrame:
-                ]
-                + [
-                    [f"{key}_pred", f"{key}_sd"]
-                    for key in self.get_keys_by_objective(
```
This should be handled in the below thread :)

gmancino replied:
Please check bofire/utils/naming_conventions.py

gmancino commented on bofire/data_models/features/categorical.py:
```diff
@@ -359,38 +361,63 @@ class CategoricalOutput(Output):
     categories: TCategoryVals
-    objective: Annotated[List[Annotated[float, Field(ge=0, le=1)]], Field(min_length=2)]
+    objective: Optional[AnyCategoricalObjective] = Field(
+        default_factory=lambda: ConstrainedCategoricalObjective(
```
Should we set desireabilities to be an array of `True`'s in this instance?

gmancino replied:
This actually ties into the above questions (i.e. regarding `None` types). I am also not sure of how to instantiate in this instance since it seems sort of "chicken-and-egg." Either way, we just need a default case here. Let me know what you prefer.

jduerholt replied:
Let us do the following, we implement the default generation into the field validator for the CategeforicalOutput:


``` python
def validate_objective(cls, objective, info):
    if objective is None:
         return ConstrainedCategoricalObjective(categories=info.data[`categories`], desirabilities = [True]*len(info.data[`categories`])]
    else:
        pass # do the validaton already implemented in the validator.
```

By this we are then setting as default that everything is allowed, which is the same as ignorig the CategoricalOutput in the optimization. 

In the attribute, we then set the folowing:

`objective: Optional[AnyCategoricalObjective] = Field(default=None, validate_default=True)`

Does this makes sense to you?

gmancino replied:
Intuitively this makes sense, however, just applying this as is actually causes the following error: FAILED tests/bofire/data_models/serialization/test_deserialization.py::test_outputs_should_be_deserializable[outputs_spec0] 

This does not make sense since these output types are Continuous. Care to comment?

gmancino commented on bofire/data_models/features/categorical.py:
```diff
@@ -359,38 +361,63 @@ class CategoricalOutput(Output):
+            Tuple[str]: Tuple of the categories
+        """
+        if self.objective.categories != self.categories:  # type: ignore
+            raise ValueError("categories must match to objective categories")
```
It will be in specs/outputs in the next commit

gmancino commented on bofire/data_models/objectives/categorical.py:
```diff
@@ -0,0 +1,83 @@
+    @field_validator(
+        "desirability",
+    )
+    def validate_categories_unique(cls, desirability: List[bool], info) -> List[bool]:
```
Makes sense, will add this. I am assuming tests should be added in `specs/objectives.py`, right?

jduerholt replied:
But you do need the validator anymore, as I introduced this global new type: https://github.com/experimental-design/bofire/blob/67f33201dee5526d1ab620a8411f83098cbce7d1/bofire/data_models/types.py#L51 which has the validation already cooked in, just set `categories: TCategoryVals` after merging main in.

gmancino replied:
So I only need the validator for desireability, correct?

gmancino commented on bofire/data_models/surrogates/api.py:
```diff
@@ -50,7 +55,8 @@
         MixedTanimotoGPSurrogate,
-        MLPEnsemble,
+        ClassificationMLPEnsemble,
+        RegressionMLPEnsemble,
```
Sure, right now, `AnyClassificatonSurrogate = ClassificationMLPEnsemble` will need to be sufficient since it is the only surrogate at this point in time

jduerholt replied:
Correct!

gmancino commented on bofire/data_models/surrogates/fully_bayesian.py:
```diff
@@ -18,3 +19,13 @@ def validate_thinning(cls, thinning, info):
+        Returns:
+            bool: True if the output type is valid for the surrogate chosen, False otherwise
+        """
+        return isinstance(my_type, ContinuousOutput)
```
Not sure what you mean here - tests in tests/bofire/data_models/specs/surrogates.py suggest this feature is working appropriately...

jduerholt replied:
Yep, it is working but it is somehow inconsistent with the `is_feature_implemented`, `is_constraint_implemented` in the strategies, there always the type of the object is passed to the is_implemented method and the actual instance.

gmancino replied:
Will update appropriately in the next commit

jduerholt commented on bofire/data_models/objectives/categorical.py:
```diff
@@ -0,0 +1,83 @@
+
+    Attributes:
+        w (float): float between zero and one for weighting the objective.
+        desirability (list): list of values of size c (c is number of categories) such that the i-th entry is in {True, False}
```
`w` is just used in case of Additive and Multiplicative Sobo strategies. What happens actually there if we use there a CategoricalOuput? I have no idea :D But lets keep the `w` and checks what happens in these strategies.

jduerholt replied:
`eta` is the steepness of the sigmoid based output constrained. In the MaxSigmoid it is actually called `steepness`, as in this categorical output constraint, the steepness is controlled totally by the classifier and its calibration, we should ignore `eta`, as we are implicitly assuming an `eta` of one in your iddentiy transformation for botorch, correct?

gmancino replied:
Correct, this should be complete now

gmancino commented on bofire/data_models/surrogates/mlp.py:
```diff
@@ -18,4 +23,44 @@ class MLPEnsemble(TrainableBotorchSurrogate):
+    scaler: ScalerEnum = ScalerEnum.NORMALIZE
+
+    @classmethod
+    def is_output_implemented(cls, my_type: Type[AnyOutput]) -> bool:
```
To clarify, your desired structure is: `Surrogate` -> `RegressionSurrogat(Surrogate)` and `ClassificationSurrogate(Surrogate)` where `is_output_implemented` is now a validator for type Continuous (for Regression) and Categorical (for Classification)? Then, each surrogate will be of type e.g. `(TrainableBotorchSurrogate,RergressionSurrogate)` for regression problems. Is this what you have in mind?

jduerholt commented on bofire/strategies/predictives/predictive.py:
```diff
@@ -102,17 +104,32 @@ def predict(self, experiments: pd.DataFrame) -> pd.DataFrame:
+                    loc=0,
+                    column=f"{feat.key}_pred",
+                    value=predictions.filter(regex=f"{feat.key}(.*)_prob")
+                    .idxmax(1)
```
This is doubled with `surrogte.py`, could we somehow also write a helper function for this and call them in both ocassions, some method called for example `postprocess_categegorical_predictions`? We could place it in `surrogate.py`. What do you think?

gmancino replied:
I will add it to the naming_conventions file. Unless you can tell me how to access the surrogates within the predictives to call a method specified within the surrogate class?

jduerholt commented on bofire/utils/naming_conventions.py:
```diff
@@ -0,0 +1,24 @@
+            for cat in outputs.get_by_key(featkey).categories  # type: ignore
+        ]
+        sd_cols = sd_cols + [
+            f"{featkey}_{cat}_sd"
```
Nice. Could you also add a test for it?

jduerholt commented on bofire/utils/naming_conventions.py:
```diff
@@ -0,0 +1,24 @@
+)
+
+
+def get_column_names(outputs) -> Tuple[List[str], List[str]]:
```
```suggestion
def get_column_names(outputs: Outputs) -> Tuple[List[str], List[str]]:
```

jduerholt replied:
And could you add a simple docstring?

jduerholt commented on tests/bofire/surrogates/test_mlp.py:
```diff
@@ -13,8 +13,8 @@
-from bofire.data_models.surrogates.api import MLPEnsemble, ScalerEnum
-from bofire.surrogates.mlp import MLP, RegressionDataSet, _MLPEnsemble, fit_mlp
+from bofire.data_models.surrogates.api import RegressionMLPEnsemble, ScalerEnum
+from bofire.surrogates.mlp import MLP, MLPDataset, _MLPEnsemble, fit_mlp
```
Can you also add a test for the Classification Ensemble?

jduerholt reviewed (commented):
Hi Gabe, 

due to the lack of time, not a full review, but a few minor things to work on. We are very close to the finish line.

Best,

Johannes

gmancino commented on tests/bofire/data_models/specs/features.py:
```diff
@@ -120,7 +123,16 @@
 )
 
-
+specs.add_valid(
```
This is done in the data_models/specs/outputs.py

jduerholt reviewed (changes requested):
Looks very good, last two things are commented. When they are finished, we can merge it in ;)

gmancino commented on bofire/data_models/objectives/categorical.py:
```diff
@@ -0,0 +1,80 @@
+
+
+class ConstrainedCategoricalObjective(
+    ConstrainedObjective, CategoricalObjective, Objective
```
Sounds good - I will have to keep `Objective` to make sure it has the appropriate functionality for being called, etc.

gmancino commented:
@jduerholt the request changes are complete :)

jduerholt reviewed (approved):
Thank you very much for all your efforts and your patience with me!
